### PR TITLE
Auto retrieve inputs of dynamic services

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

when a computational service linked to a dynamic service is uploading data, the connected dynamic services shall automatically retrieve their input data, thus making it available in the service.

**NOTE1:** for auto-retrieve to work with 2 dynamic services, another PR will be opened that will bring the checksum in
**NOTE2:** for auto-retrieve with frontend nodes, another PR will be opened that will call auto-retrieve when the project gets saved

auto-retrieve

**Details:**
- Refactored postgres notification listener code in webserver
- notification from postgres now triggers for any kind of class of task (provided the comp_tasks outputs/state column changed) --> **migration of postgres**
- notification from postgres now passes which column triggered the change with the notification payload -> **migration of postgres**
- added test_comp_tasks in postgres library
- added test for listening task in webserver
- director-v2 now provide the service "retrieve" API to allow backend call to the retrieve endpoint on dynamic services
- when the outputs in the comp_tasks table changes, the webserver calls the director-v2 to trigger the retrieve endpoint on possible following dynamic services, thus they "auto-retrieve" the data .
## Related issue number

[#856](https://github.com/ITISFoundation/osparc-simcore/issues/856)

## Checklist

- [x] Did you change any service's API? Then make sure to bundle document and upgrade version (``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [ ] Documentation reflects the changes
- [ ] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

File: packages/postgres-database/tests/test_comp_tasks.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    updated_output = {"some new stuff": "it is newer"}
+    await _update_comp_task_with(db_connection, task, outputs=updated_output)
+    await _update_comp_task_with(db_connection, task, outputs=updated_output)
+    tasks = await _assert_notification_queue_status(db_notification_queue, 1)
+    assert tasks[0]["changes"] == ["outputs"]
+    assert (
+        tasks[0]["data"]["outputs"] == updated_output
+    ), f"the data received from the database is {tasks[0]}, expected new output is {updated_output}"
+
+    # updating twice with different stuff comes out in FIFO order
```
Comment (sanderegg) [inline packages/postgres-database/tests/test_comp_tasks.py:125]:
```suggestion
    # updating a number of times with different stuff comes out in FIFO order
```

**Review by odeimaiz — APPROVED** [review]
👌

File: services/web/server/src/simcore_service_webserver/director_v2.py [hunk 2]
```diff
@@ -176,6 +175,23 @@ async def stop_pipeline(request: web.Request) -> web.Response:
...
+    backend_url = URL(
+        f"{director2_settings.endpoint}/dynamic_services/{service_uuid}:retrieve"
+    )
+    body = {"port_keys": port_keys}
+
+    try:
+        # request to director-v2
+        await _request_director_v2(app, "POST", backend_url, data=body)
+    except _DirectorServiceError:
+        pass
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/director_v2.py:192]: is this desired?

**Review by GitHK — APPROVED** [review]
Great, I only have a few questions.

**Review by KZzizzle — APPROVED** [review]

File: packages/postgres-database/tests/test_comp_tasks.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    tasks = await _assert_notification_queue_status(db_notification_queue, 1)
+    assert tasks[0]["changes"] == ["outputs", "state"]
+    assert (
+        tasks[0]["data"]["outputs"] == updated_output
+    ), f"the data received from the database is {tasks[0]}, expected new output is {updated_output}"
+
+    # setting the exact same data twice triggers only ONCE
+    updated_output = {"some new stuff": "it is newer"}
+    await _update_comp_task_with(db_connection, task, outputs=updated_output)
+    await _update_comp_task_with(db_connection, task, outputs=updated_output)
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/postgres-database/tests/test_comp_tasks.py:118]: the behavior is dictated by how the trigger is defined within the postgres server. In which case it only triggers when the output or the state changes (IS DISTINCT OF) you can find that [here](https://github.com/sanderegg/osparc-simcore/blob/221edcd9728e83ffee6900df3948722157d706e4/packages%2Fpostgres-database%2Fsrc%2Fsimcore_postgres_database%2Fmodels%2Fcomp_tasks.py#L55-L64)

File: services/director-v2/src/simcore_service_director_v2/utils/clients.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+def unenvelope_or_raise_error(resp: httpx.Response) -> Dict:
+    """
+    Director responses are enveloped
+    If successful response, we un-envelop it and return data as a dict
+    If error, it raise an HTTPException
+    """
+    body = resp.json()
+
+    assert "data" in body or "error" in body  # nosec
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/utils/clients.py:20]: this is called defensive programming, you can ask @pcrespov about it. it is stripped out in production and will not fail in that case.

File: services/web/server/src/simcore_service_webserver/director_v2.py (hunk 2 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/director_v2.py:192]: good one, I added a warning now to log in case of issues.

**Pull request merged**